### PR TITLE
Add FreeBSD x86_64 debug tests

### PR DESCRIPTION
**Pull request by XVilka**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Fixed register profile for freebsd-x64
- Fixed fetching of maps using sysctl .
- Base address is fetched from debug maps which have at least read permission (used for non linux and non windows system)
Adds basic FreeBSD x86_64 debug tests.

Requires https://github.com/rizinorg/rizin-testbins/pull/101

**Test plan**

FreeBSD CI is green

**Closing issues**

Partially addresses https://github.com/rizinorg/rizin/issues/3072

### CODE REVIEW

**Comment by XVilka** [comment]
@arrowd, you might be interested in this one, by the way. Feel free to send a new PR if you want.

**Comment by XVilka** [comment]
@kazarmy as you can see, even FreeBSD debugging doesn't work:
```
[XX] db/archos/freebsd-x64/dbg write gpr
RZ_NOPLUGINS=1 /home/build/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -d -qc 'db @ main + 64
dc
dr rax
dr rax = 1337
dc
' bins/elf/calculate-freebsd-x64
-- stdout
--- expected
+++ actual
@@ -1,2 +1,2 @@
-rax = 0x000000000000002a
-Result = 1337
+Result = 42
+Debugging is not enabled. Run ood?

-- stderr
--- expected
+++ actual
@@ -1,4 +1,27 @@
-hit breakpoint at: f70
+Process 
+with 
+PID 
+387 
+child 
+exited 
+with 
+status 
+Process 
+finished
 
-Process finished
 
+ERROR: 
+Failed 
+to 
+read 
+ERROR: 
+No 
+such 
+register 
+or 
+register 
+type: 
+ERROR: 
+Failed 
+to 

```
https://builds.sr.ht/~thestr4ng3r/job/1239045

**Comment by kazarmy** [comment]
Well that's one reason why no testing was done in #4524 - debugging might have been broken well before the build itself broke

**Review by wargio — APPROVED** [review]

**Comment by well-mannered-goat** [comment]
Wrote some changes description in the PR

Also in FreeBSD `sysctl` is used to fetch maps first, because `procfs` is not mounted on `\proc\` hence no pid directories are there by default. Also the parsing logic is kind of flawed because of different map format (after mounting procfs)
They aren't fetched from `/proc/` (its the second option if `sysctl` does not work), so I thought this is redundant code and with bugs. Should I remove it?

**Review by notxvilka — APPROVED** [review]

**Comment by notxvilka** [comment]
> Wrote some changes description in the PR
> 
> Also in FreeBSD `sysctl` is used to fetch maps first, because `procfs` is not mounted on `\proc\` hence no pid directories are there by default. Also the parsing logic is kind of flawed because of different map format (after mounting procfs) They aren't fetched from `/proc/` (its the second option if `sysctl` does not work), so I thought this is redundant code and with bugs. Should I remove it?

It's better to keep but open an issue about it, then we can decide what to do with it in a separate PR.

File: librz/core/cfile.c [hunk 1]
```diff
@@ -586,8 +586,9 @@ static ut64 get_base_from_maps(RzCore *core, const char *file) {
 	ut64 b = 0LL;
 
 	rz_debug_map_sync(core->dbg); // update process memory maps
+	// check for the first entry which has read permission
 	rz_list_foreach (core->dbg->maps, iter, map) {
-		if ((map->perm & 5) == 5) {
+		if (map->perm & 4) {
```
Comment (wargio) [inline librz/core/cfile.c]: Please use macros for constants. Also there are already macros for the perms.

File: librz/core/cfile.c [hunk 2]
```diff
@@ -604,8 +605,9 @@ static ut64 get_base_from_maps(RzCore *core, const char *file) {
 		}
 	}
 	// fallback resolution copied from cmd_debug.c:rz_debug_get_baddr
+	// check for the first entry which has read permission
 	rz_list_foreach (core->dbg->maps, iter, map) {
-		if (map->perm == 5) { // r-x
+		if (map->perm & 4) { // r--
```
Comment (wargio) [inline librz/core/cfile.c]: Same here

File: librz/core/cfile.c (hunk 1 above)
Comment (wargio) [reply librz/core/cfile.c]: Add also a comment explaining why that value is needed

File: librz/debug/p/native/bsd/bsd_debug.c [hunk 3]
```diff
@@ -368,6 +360,9 @@ RzList *bsd_native_sysctl_map(RzDebug *dbg) {
 	struct kinfo_vmentry *kve;
 	RzList *list = NULL;
 	RzDebugMap *map;
+	char *name;
```
Comment (wargio) [inline librz/debug/p/native/bsd/bsd_debug.c]: Initialize these new values to null/0

File: librz/debug/p/native/kfbsd.c [hunk 4]
```diff
@@ -40,9 +40,35 @@ static bool rz_debug_native_step(RzDebug *dbg) {
 	return true;
 }
 
+int match_pid(const void *pid_o, const void *th_o, void *user) {
```
Comment (wargio) [inline librz/debug/p/native/kfbsd.c:43]: Is this a boolean,? If yes use true / false

File: librz/debug/p/native/kfbsd.c [hunk 5]
```diff
@@ -40,9 +40,35 @@ static bool rz_debug_native_step(RzDebug *dbg) {
...
+	int pid = *(int *)pid_o;
+	RzDebugPid *th = (RzDebugPid *)th_o;
+	return (pid == th->pid) ? 0 : 1;
```
Comment (wargio) [inline librz/debug/p/native/kfbsd.c]: Why not just `return pid != th->pid` ?

File: librz/debug/p/native/kfbsd.c [hunk 6]
```diff
@@ -148,6 +170,7 @@ static RzDebugReasonType rz_debug_native_wait(RzDebug *dbg, int pid) {
 	}
 	dbg->reason.tid = pid;
 	dbg->reason.type = reason;
+	eprintf("Here is your reason for wait %d\n", reason);
```
Comment (wargio) [inline librz/debug/p/native/kfbsd.c]: Warn? Also can you rephrase this to be more professional ?

**Review by wargio — CHANGES_REQUESTED** [review]
Nice improvement

File: librz/core/cfile.c (hunk 1 above)
Comment (well-mannered-goat) [reply librz/core/cfile.c]: Yep, forgot about that
Comment as in? initially it fetched base from map which had r-x permissions which would shift up the symbols address to higher address. This is just fetches the base of the whole process.

File: librz/debug/p/native/kfbsd.c (hunk 4 above)
Comment (well-mannered-goat) [reply librz/debug/p/native/kfbsd.c:43]: Nope, its used as comparator in rz_list_find

File: librz/debug/p/native/kfbsd.c (hunk 6 above)
Comment (well-mannered-goat) [reply librz/debug/p/native/kfbsd.c]: Oops, sorry about that. Used it for debugging in VMs :sweat_smile: . Cant really use debuggers inside VM

**Comment by well-mannered-goat** [comment]
Forgot about the step over test, will also add it

**Comment by well-mannered-goat** [comment]
> > Wrote some changes description in the PR
> > Also in FreeBSD `sysctl` is used to fetch maps first, because `procfs` is not mounted on `\proc\` hence no pid directories are there by default. Also the parsing logic is kind of flawed because of different map format (after mounting procfs) They aren't fetched from `/proc/` (its the second option if `sysctl` does not work), so I thought this is redundant code and with bugs. Should I remove it?
> 
> It's better to keep but open an issue about it, then we can decide what to do with it in a separate PR.

I'll just open up an issue then.

File: librz/core/cfile.c (hunk 1 above)
Comment (wargio) [reply librz/core/cfile.c:591]: like in the line below. `// r--`

**Review by wargio — APPROVED** [review]

File: librz/debug/p/native/kfbsd.c [hunk 7]
```diff
@@ -99,17 +122,16 @@ static RzDebugReasonType rz_debug_native_wait(RzDebug *dbg, int pid) {
 	/* we don't know what to do yet, let's try harder to figure it out. */
 	if (reason == RZ_DEBUG_REASON_UNKNOWN) {
 		if (WIFEXITED(status)) {
-			eprintf("child exited with status %d\n", WEXITSTATUS(status));
+			if (dbg->pid == pid) {
+				RZ_LOG_WARN("(%d) Process exited with status=0x%x\n", pid, WEXITSTATUS(status));
```
Comment (Rot127) [inline librz/debug/p/native/kfbsd.c]:
```suggestion
				RZ_LOG_WARN("(%" PFMT32d ") Process exited with status=0x%x\n", pid, WEXITSTATUS(status));
```

File: librz/debug/p/native/kfbsd.c [hunk 8]
```diff
@@ -99,17 +122,16 @@ static RzDebugReasonType rz_debug_native_wait(RzDebug *dbg, int pid) {
...
+			} else {
+				RZ_LOG_WARN("(%d) Thread exited with status=0x%x\n", pid, WEXITSTATUS(status));
```
Comment (Rot127) [inline librz/debug/p/native/kfbsd.c]:
```suggestion
				RZ_LOG_WARN("(%" PFMT32d ") Thread exited with status=0x%x\n", pid, WEXITSTATUS(status));
```

**Review by Rot127 — APPROVED** [review]
